LSM: Don't overwrite non-owned memory with undefined in Groove

ifreund opened the pull request:
Two other cleanup commits are included as well, see the commit messages for detail.

ifreund committed:
LSM: clean up loop in Groove.PrefetchContext.start_workers()

ifreund committed:
LSM: Don't overwrite non-owned memory with undefined in Groove

Currently it's possible for the prefetch code to overwrite the memory
previously occupied by one of the prefetch workers with undefined after
the prefetch has finished and the callback has already been called.

This means that if the PrefetchContext memory is re-used in the callback
we overwrite memory now used for something else with undefined.

ifreund committed:
LSM: Reduce dimensionality of prefetch code in Groove

jorangreef commented on src/lsm/groove.zig:
```diff
@@ -642,34 +642,32 @@ pub fn GrooveType(
+                    // worker_finished() as the memory may already be used for something else.
+                    const context = worker.context;
+                    worker.* = undefined;
+                    context.worker_finished();
```
Let's go one step further still and not let any descendant node in a call graph erase its own memory at all.

i.e. Just do `worker.context.worker_finished();` here with no `worker.* = undefined;`.

Rationale:

1. The process that kicks off all the concurrent workers is responsible for initializing their memory, however, at present, that same memory is being set to `undefined` not at the same level of concurrency control, but one node deeper in the call graph. This is rather the responsibility of whoever owns the memory, so as not to violate encapsulation.

2. This is data plane—we don't want to do memsets for every id/object. For example, in future, an object may be 1024 bytes or 16384 bytes etc.

3. It's going to simplify callback call sites in general. From 3 lines down to 1. This “copy callback“ and then set to undefined pattern can be unwieldy and subtly tricky to get right. I want us to start moving away from it. For example, on line 632, we have another potential future issue, where `finish()` sets `context.* = undefined` before calling `callback(context)` passing that same `context` which is now `undefined`—it may be that in future `callback()` still wants to access something of `context`, rather than only using it for `@fieldParentPtr()`, and this is especially likely if the memory is managed at a higher layer.

ifreund replied:
Yeah, I agree that this pattern doesn't seem to be worth the tradeoff in practice. Added a commit removing all instances of it across our code.

(Aside: what do you think of the "style" commit message prefix for this kind of thing?)

jorangreef replied:
Great, thanks for removing all the other instances.

I would rather not introduce commit message prefixes beyond the functional namespaces (e.g. `LSM`, `VSR` etc). Code is style/design (safety/performance/experience) and design is code. It's hard to separate those things. They're so closely joined into the conceptual integrity of what makes TB, TB.

ifreund replied:
Makes sense to me, switched to `LSM,VSR` for the prefix :+1:

jorangreef reviewed (changes requested):
Nice fix @ifreund. Two changes to eliminate the “set to `undefined`“ by the user of the memory (rather than by the owner of the memory) and then it's good to go.

ifreund committed:
LSM,VSR: Don't set context to undefined before calling callbacks

This pattern that attempts to make code safer is causing more issues
than it is worth. It's difficult to get correct, especially in less
trivial cases such as the Prefetch code in Groove where we recently
fixed a bug caused by getting this wrong that took quite a while
to find.

It also hasn't helped us find any bugs yet, the many assertions seem to
cover the cases where it might help already.
